Conditions 671 Update Feature Toggle for Platform ArrayField and Autocomplete Progressive Rollout

williamphelps13 opened the pull request:
Demo: https://dsva.slack.com/archives/C04KW0B46N5/p1738186910711159?thread_ts=1738184502.092349&cid=C04KW0B46N5

## Summary
- Summarize the changes that have been made to the platform: 
  - Storing the feature toggle in sessionStorage does not work in production because by the time the toggle value is stored and then retrieved the form has already loaded. This approach stores the feature toggle in the formData and uses depends to check that formData and conditionally show the correct content.
  - Changes by file:
    - [all-claims/Form526EZApp.jsx](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-7160eda0f93aa9a0dc5391889d45cf5fd5ab5c33ecd281c9c163aa2f5bc452bd)
    - [all-claims/components/NewDisability.jsx](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-e9bef5491f8fe39105fbb335720733b04345ab3914d135e6e727937249dcbe4c)
    - [all-claims/config/form.js](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-d31fa75e078e845b8d2489756cb223670afb3342c26d135c12984d098e6cd1c2)
    - [all-claims/constants.js](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-baae3f550e35220e2d29206d29c8649813021f096b61876c98a6bb9fbe354b1e)
    - [all-claims/content/addDisabilities.jsx](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-4631fe5289b967d896025bdd9b6777a0845b7b1c2b47801e70a04c464553c715)
    - [all-claims/pages/addDisabilities.js](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-7a9dd7c424ab9980579d6167e1db4743e1b1f20b532d65b00e9eaaee2c20a75f)
    - [all-claims/pages/addDisabilitiesPrevious.js](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-a159341f0bfefec740ff0bb4d40b0b007358fe3b2f724bcac1befc7d2153c3fd)
    - [all-claims/pages/index.js](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-c7e68a1912db01b3d1b29cec11bae0e0702fa186c5c81c5fa9644a642f605331)
    - [all-claims/tests/cypress.helpers.js](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-e884ea03b47a9baa004b871fd4f054b02ada591cbcc9e0f5435d590da463a46d)
    - [all-claims/tests/fixtures/mocks/feature-toggles.json](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-bc20074b2909c8e6fe5b47abe06c2db592dcc5ec48860a403f62c7bbcb6df2c0)
    - [all-claims/tests/pages/addDisabilities.unit.spec.jsx](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-deac3e68494bc22ca134579b7caef7c1dc543e5b6a5405a84b44433351709c0c)
    - [all-claims/tests/pages/addDisabilitiesPrevious.unit.spec.jsx](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-f084c44b0df802c6c9b3306e07e9a67805cd955a522a4db5c05e1b271d733ebe)
    - [all-claims/tests/property-names.unit.spec.js](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-5d8e118aa326d22f0e6a40bc92d10867a7a72f926af906f89acaddf21214799e)
    - [all-claims/utils/index.jsx](https://github.com/department-of-veterans-affairs/vets-website/pull/34038/files#diff-5732ef633479668ab61be2210a1fc0a2695f68ff3c1d544c686ec634b96862f6)
      - new content is at add-3
        - This is in alignment with the [advice from Information Architecture](https://dsva.slack.com/archives/C01K37HRUAH/p1737485451871039?thread_ts=1721329260.337369&cid=C01K37HRUAH). First they advised to just alter the page the user is on: "Change the URL slug for the exact page, not a parent page - so if you are changing the page that lives at "/add" in this URL `/benefit/apply-form123/dependents/add`, then change the "/add" slug, not the "/dependents/ slug"." Second they advised to "Avoid any words that could create confusion or alarm with the user that they may not be on the right page.  In other words, I would stay away from using terms like "old" or "archive", and instead use something that it less noticeable like just appending a number to the URL." The content of this page is the 3rd major version of the content (if we call the longstanding content v1 and the Combobox updated v2) hence the "/add-3". Either way this is just a temporary URL that will be removed in a short time (2 weeks at most) and will aim to only have enough users go on the URL to determine the new content is working well.
- Which team do you work for, does your team own the maintenance of this component? Conditions Team and Yes
- If using a flipper, what is the end date of the flipper being required/success criteria being targeted: After progressive rollout using the code from this PR the next PR will remove the feature toggle

## Related issue
- [Replace Combobox and All Claims ArrayField with Autocomplete and Platform ArrayField #671](https://github.com/department-of-veterans-affairs/vagov-claim-classification/issues/671)

## Testing done

### Describe the tests completed and the results: 
Note: The URL has since changed from these recordings. Previously the plan was to put the old content on `/add-2` and new content on`/add`. Now the old content will remain on `/add` and the new content will instead go on `/add-3` (3 because it is the major version of the page content). This will allow a slower rollout with less users going onto a URL that will not be tracked by GA longterm.
1.  `Feature Toggle = Disabled` and User is working on form --> `Feature Toggle = Enabled`:  https://dsva.slack.com/archives/C04QKNK4J0H/p1736886252029299?thread_ts=1736867918.896029&cid=C04QKNK4J0H
2.  `Feature Toggle = Enabled` and User is working on form --> `Feature Toggle = Disabled`: https://dsva.slack.com/archives/C04QKNK4J0H/p1736886276091769?thread_ts=1736867918.896029&cid=C04QKNK4J0H
3.  `Feature Toggle = Disabled` and User saved form for later --> `Feature Toggle = Enabled`; User starts form again: https://dsva.slack.com/archives/C04QKNK4J0H/p1736886303464249?thread_ts=1736867918.896029&cid=C04QKNK4J0H
4.  `Feature Toggle = Enabled` and User saved form for later --> `Feature Toggle = Disabled`; User starts form again: https://dsva.slack.com/archives/C04QKNK4J0H/p1736886318550179?thread_ts=1736867918.896029&cid=C04QKNK4J0H
5.  `Feature Toggle = Enabled` and User saved form for later **_while on different page then the addDisabilities page_**--> `Feature Toggle = Disabled`; User starts form again: https://dsva.slack.com/archives/C04QKNK4J0H/p1736886403797389?thread_ts=1736867918.896029&cid=C04QKNK4J0H
 
## Screenshots
#### Page Load
<img width="1685" alt="Screenshot 2024-12-17 at 7 43 25 AM" src="https://github.com/user-attachments/assets/897231fb-68c3-48de-95be-34c0c009e8f8" />

#### Click Continue with No Conditions
<img width="1681" alt="Screenshot 2024-12-17 at 7 47 43 AM" src="https://github.com/user-attachments/assets/8f745bcd-177a-41b3-8507-4aa9acc35254" />

#### Adding Condition to Autocomplete
<img width="1686" alt="Screenshot 2024-12-17 at 7 43 59 AM" src="https://github.com/user-attachments/assets/980b444f-5d24-4853-82cb-d7259f7d2edc" />

#### Condition Saved
<img width="1685" alt="Screenshot 2024-12-17 at 7 46 07 AM" src="https://github.com/user-attachments/assets/14e522a9-663a-4f7d-ae8e-03c685679665" />

#### Save Condition with No Text in Autocomplete
<img width="1684" alt="Screenshot 2024-12-17 at 7 46 52 AM" src="https://github.com/user-attachments/assets/d021adba-aceb-481f-96e6-f119c6395b84" />

#### Click Remove
![Screenshot 2024-12-17 at 7 47 03 AM](https://github.com/user-attachments/assets/9bc137a1-570e-4360-8f7f-e807f886acf8)

#### Click Add Another Condition while Editing A Condition (Both Then Autosave) then Click Edit in the Above Condition
<img width="1686" alt="Screenshot 2024-12-17 at 7 56 58 AM" src="https://github.com/user-attachments/assets/16c65aa6-b2ea-41e4-96e4-832eda74f7e8" />

## Acceptance criteria

### Quality Assurance & Testing

- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x] Linting warnings have been addressed
- [x] Documentation has been updated ([link to documentation](#) \*if necessary)
- [x] Screenshot of the developed feature is added
- [x] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/collaboration-cycle/prepare-for-an-accessibility-staging-review) has been performed

### Error Handling

- [x] Browser console contains no warnings or errors.
- [x] Events are being sent to the appropriate logging solution
- [x] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [x] Did you login to a local build and verify all authenticated routes work as expected with a test user

## Next steps

### Removing Feature Toggle
There are 3 scenarios I can think of for the interaction of users with this progressive rollout and how they are effected by removing the Feature Toggle:
  
 1. They saved the form on the '/new-disabilities/add' page back in November, they don't login during the progressive rollout --> No effect on user experience
 2.  They saved the form on the '/new-disabilities/add' page during the rollout (they were NOT part of the percent it was rolled out to) --> No effect on user experience
 3. They saved the form on the '/new-disabilities/add-3' page during the rollout (they were part of the percent it was rolled out to) -->
     1. --> we move the new content on '/new-disabilities/add-3' to '/new-disabilities/add' and delete all code from progressive rollout --> Save in Progress would try to route them to '/new-disabilities/add-3' but because that URL and  onFormLoaded function was deleted they would be routed to the first page of the form '/contact-information'; SMALL  impact to user experience
    2. --> we move the new content on '/new-disabilities/add-3' to '/new-disabilities/add' and add an *UPDATED* version of the onFormLoaded function (see below) --> Save in Progress would try to route them to '/new-disabilities/add-3' but because that URL was removed the onFormLoaded function would route them to '/new-disabilities/add'; NO impact to user experience --> Remove this after no more users have Save in Progress at '/new-disabilities/add-3',
    ```
     /**
     * @param {String} returnUrl - URL of last saved page
     * @param {Object} router - React router
     */
    export const onFormLoaded = props => {
      const { returnUrl, router } = props;

      if (returnUrl === '/new-disabilities/add-3') {
        router?.push('/new-disabilities/add');
      } else {
        router?.push(returnUrl);
      }
    };
    ```

### Rollout Plan
MERGE TO PROD DAY: 1/30/25
ROLLOUT DAY: 2/3/25
ROLLOUT DAY: 2/3/25 - 2/8/25

- [ ] Merge PR on day before MERGE TO PROD DAY after that day's prod deployment
- [ ] Run bug bash on day before MERGE TO PROD DAY
  - If there is an issue we can revert that commit or fix and prod will not be effected
- [ ] If all are satisfied, Prod Deployment could continue on MERGE TO PROD DAY 1:00pm EST 
  - All users will by default be routed to old content at `/add` so no change should occur immediately on prod deployment, but will monitor based on Monitoring Plan
- [ ] Wait until 2am EST on ROLLOUT DAY to increase feature toggle to 5%
- [ ] ROLLOUT DAY MONITORING: Look closely at `/add-3` sessions and ensure they are occurring AND compare these `add-3` sessions to the count of users that have a session on `/follow-up` page with `/add-3` being the referrer. As noted in a [thread with analytics](https://dsva.slack.com/archives/CKCRMLGAC/p1738080734004479?thread_ts=1738078674.366029&cid=CKCRMLGAC) real time GA analysis is not great so we cannot look for percentages  but only dramatic decreases
  - Ensure there are continued sessions on `/add-3` and `/add`
  - Ensure users are continuing from `/add-3` and `/add` to `/follow-up`
- [ ] ROLLOUT WEEK MONITORING: Ensure the ~82% trend of `/add` + `/add-3` sessions to `/follow-up` continue in this completion rate spreadsheet
- [ ] When new content is working great for users we will merge a PR to do the following:
  - Move the new content from the temporary url (`/new-disabilities/add-3`) to the standard url (`/new-disabilities/add`) and remove the temporary url (`/new-disabilities/add-3`) from the form
  - Remove the ComboBox code, feature toggle code, and all other code that is from the "old content"
  - Update the `onFormLoaded` function to simply route users that saved on the temporary url to the standard url (no longer incorporating any feature toggle logic)
  - Update the `onFormLoaded` TODO to include the date in which that function can be completely removed (1 year from the end of the rollout)  (Note: even if that function were removed early it would just reroute any users that saved on the temporary url to the first page of the form)

williamphelps13 committed:
Update feature toggle to be stored in formData rather than sessionStorage

Storing the feature toggle in sessionStorage does not work in production because by the time the toggle value is stored and then retrieved the form has already loaded. This approach stores the feature toggle in the formData and uses depends to check that formData and conditionally show the correct content.

williamphelps13 committed:
Update Autocomplete addDisabilities tests for Platform ArrayField

Platform ArrayField 'Edit' button is a va-button whereas All Claims ArrayField 'Edit' button is not. va-button requires a test util since it is a web component and text cannot be retrieved with getByText.

williamphelps13 committed:
Add check fo feature toggle being undefined to fix failing tests

williamphelps13 committed:
Fix duplicate property names test by adding to the ignore array

williamphelps13 committed:
Update e2e testing to update var name & test with toggle on

williamphelps13 committed:
Fix storing of feature toggle in formData and onFormLoad

Aligns with the implementation found in src/applications/appeals/995/containers/App.jsx. Now the feature toggle setup successfully allows rolling out and rolling back the feature toggle. The only remaining issue is that if they are currently in the form when the feature toggle is toggled they will get a error message and need to refresh their browser.

williamphelps13 committed:
Resolve failing App tests by adding formData to check and to dep array

williamphelps13 committed:
Merge branch 'main' into conditions-671-update-feature-toggle-handling-to-allow-platform-array-field-progressive-rollout

bacitracin commented on src/applications/disability-benefits/all-claims/Form526EZApp.jsx:
```diff
@@ -282,15 +295,21 @@ Form526Entry.propTypes = {
 
 const mapStateToProps = state => ({
   accountUuid: state?.user?.profile?.accountUuid,
+  formData: state.form.data,
```
Should this also be `state?.form?.data` ?

williamphelps13 committed:
Fix tests from merging main with ArrayField VaButton update

This PR updated the 'update' and 'remove' buttons to VaButtons: https://github.com/department-of-veterans-affairs/vets-website/pull/34024/files.

bacitracin commented on src/applications/disability-benefits/all-claims/components/NewDisability.jsx:
```diff
@@ -2,17 +2,11 @@ import React from 'react';
-        showEnhancement ? 'vads-u-margin-left--1' : 'vads-u-flex--fill'
-      } vads-u-padding-right--2 word-break capitalize-first-letter`}
-    >
+    // TODO: Remove vads-u-flex--fill when allClaimsAddDisabilitiesEnhancement feature flag is removed
```
Just curious, what's the diff here? Platform array field has diff margins or something?

bacitracin replied:
Thank you!

williamphelps13 committed:
Fix e2e tests from merging main with ArrayField VaButton update

This PR updated the 'update' and 'remove' buttons to VaButtons: https://github.com/department-of-veterans-affairs/vets-website/pull/34024/files\.

williamphelps13 committed:
Fix null checks for formData in mapStateToProps to prevent potential runtime errors

From PR feedback: https://github.com/department-of-veterans-affairs/vets-website/pull/34038#discussion_r1916958897

christinec-fftc commented on src/applications/disability-benefits/all-claims/utils/index.jsx:
```diff
@@ -874,3 +885,27 @@ export const formatFullName = (fullName = {}) => {
   environment.isDev() || environment.isLocalhost() || environment.isStaging();
+
+/**
+ * // TODO: Remove when allClaimsAddDisabilitiesEnhancement feature flag is removed. This feature flag can only be removed when no more save in progress forms are left on the new-disabilities-previous/add page.
```
Looks great! Just a couple items
1. It might be easier to find all these todo's later if a ticket number is listed with each TODO comment
2. Is it true the feature flag can only be removed when there are no more IPF's on the old page? So like up to one year post launch?

christinec-fftc replied:
Ticket # preferred, but I admit I haven't even always followed this 🙃 It's nice when the repo/linter enforces this.

Thanks for the detail. That is a lot to grok. My only concern for 2.ii is just how challenging it might be to identify when there are no more returnUrl's of a given value. That seems like a moving target until the page is removed from the form config, and then from their you have to wait til any IPF's with that return url update?

bacitracin commented:
It looks good but you have some warnings around the save in progress so be sure to check that out https://github.com/department-of-veterans-affairs/vets-website/actions/runs/12793955645/job/35668463319

williamphelps13 committed:
Merge branch 'main' into conditions-671-update-feature-toggle-handling-to-allow-platform-array-field-progressive-rollout

williamphelps13 committed:
Update addDisabilitiesPrevious page path to be less  confusing to users

This is in alignment with the advice from Information Architecture. First they advised to just alter the page the user is on: "Change the URL slug for the exact page, not a parent page - so if you are changing the page that lives at "add" in this URL /benefit/apply-form123/dependents/add, then change the /add slug, not the /dependents/ slug.". Second they advised to "Avoid any words that could create confusion or alarm with the user that they may not be on the right page.  In other words, I would stay away from using terms like "old" or "archive", and instead use something that it less noticeable like just appending a number to the URL." The content of this page is the 2nd version of the content hence the /add-2.

williamphelps13 committed:
Switch new content path to /add-3 and old content path to /add

This will allow us to rollout the feature more slowly as 95% of users will go to /add and only 5% to /add-3 for a extended time to allow seeing forms get fully submitted, and in the meantime the Google Analytics will be only slightly disrupted. The add-3 references that this is the 3rd version of the page. When it is assured that /add-3 content is working well a change will be merged to move the content from /add-3 to /add.

williamphelps13 committed:
Update cypress helper in alignment with url path update

williamphelps13 committed:
Add issue url to TODO and update onFormLoaded TODO from PR feedback

williamphelps13 committed:
Merge branch 'main' into conditions-671-update-feature-toggle-handling-to-allow-platform-array-field-progressive-rollout

bacitracin reviewed (dismissed):
👍

williamphelps13 committed:
Merge branch 'main' into conditions-671-update-feature-toggle-handling-to-allow-platform-array-field-progressive-rollout
